### ISSUE TITLE
[QST] Any rules to follow when set instruction shape for GemmTensorOp?

### ISSUE THREAD

**Issue by YJHMITWEB** [NVIDIA/cutlass#1341]
**What is your question?**
Hi, I am looking at `example_19_tensorop_canonical`, and I have two questions. **(I am using A100 GPUs with CUDA 12.2)**

**Q1.**
I understand this is a warp MMA example. And in the following code:
```
  using GemmTensorOp = cutlass::gemm::warp::GemmTensorOp<
    cutlass::gemm::GemmShape<kM, kN, kK>,
    cutlass::gemm::GemmShape<8, 8, 4>,
    double,                             // Data type of A elements
    cutlass::layout::RowMajor,          // Layout of A matrix
    double,                             // Data type of B elements
    cutlass::layout::ColumnMajor,       // Layout of B matrix
    double,                             // Data type of C elements
    cutlass::layout::RowMajor,          // Layout of C matrix
    double                              // Scalar type of alpha and beta
  >;
```
It uses `m8n8k8` as the instruction shape, and I can find a corresponding `mma` instruction in PTX, as follows:

`Double precision floating point type:`
```
mma.sync.aligned.shape.row.col.f64.f64.f64.f64 d, a, b, c;

.shape   = {.m8n84, .m16n8k4, .m16n8k8, .m16n8k16};
```
Since this example uses `double`, I see there are other instruction shapes available, e.g. `m16n8k4, .m16n8k8, .m16n8k16`. 

However, if I change `cutlass::gemm::GemmShape<8, 8, 4>` into, for example, `cutlass::gemm::GemmShape<16, 8, 8>` , it raises the following error:

```
void cutlass::arch::Mma<cutlass::gemm::GemmShape<16, 8, 8>, 32, double, cutlass::layout::RowMajor, double, cutlass::layout::ColumnMajor, double, cutlass::layout::RowMajor, cutlass::arch::OpMultiplyAdd>::operator()(cutlass::Array<double, 4, true> &, const cutlass::Array<double, 4, true> &, const cutlass::Array<double, 2, true> &, const cutlass::Array<double, 4, true> &) const not implemented
void cutlass::arch::Mma<cutlass::gemm::GemmShape<16, 8, 8>, 32, double, cutlass::layout::RowMajor, double, cutlass::layout::ColumnMajor, double, cutlass::layout::RowMajor, cutlass::arch::OpMultiplyAdd>::operator()(cutlass::Array<double, 4, true> &, const cutlass::Array<double, 4, true> &, const cutlass::Array<double, 2, true> &, const cutlass::Array<double, 4, true> &) const not implemented
void cutlass::arch::Mma<cutlass::gemm::GemmShape<16, 8, 8>, 32, double, cutlass::layout::RowMajor, double, cutlass::layout::ColumnMajor, double, cutlass::layout::RowMajor, cutlass::arch::OpMultiplyAdd>::operator()(cutlass::Array<double, 4, true> &, const cutlass::Array<double, 4, true> &, const cutlass::Array<double, 2, true> &, const cutlass::Array<double, 4, true> &) const not implemented
void cutlass::arch::Mma<cutlass::gemm::GemmShape<16, 8, 8>, 32, double, cutlass::layout::RowMajor, double, cutlass::layout::ColumnMajor, double, cutlass::layout::RowMajor, cutlass::arch::OpMultiplyAdd>::operator()(cutlass::Array<double, 4, true> &, const cutlass::Array<double, 4, true> &, const cutlass::Array<double, 2, true> &, const cutlass::Array<double, 4, true> &) const not implemented
void cutlass::arch::Mma<cutlass::gemm::GemmShape<16, 8, 8>, 32, double, cutlass::layout::RowMajor, double, cutlass::layout::ColumnMajor, double, cutlass::layout::RowMajor, cutlass::arch::OpMultiplyAdd>::operator()(cutlass::Array<double, 4, true> &, const cutlass::Array<double, 4, true> &, const cutlass::Array<double, 2, true> &, const cutlass::Array<double, 4, true> &) const not implemented
void cutlass::arch::Mma<cutlass::gemm::GemmShape<16, 8, 8>, 32, double, cutlass::layout::RowMajor, double, cutlass::layout::ColumnMajor, double, cutlass::layout::RowMajor, cutlass::arch::OpMultiplyAdd>::operator()(cutlass::Array<double, 4, true> &, const cutlass::Array<double, 4, true> &, const cutlass::Array<double, 2, true> &, const cutlass::Array<double, 4, true> &) const not implemented
void cutlass::arch::Mma<cutlass::gemm::GemmShape<16, 8, 8>, 32, double, cutlass::layout::RowMajor, double, cutlass::layout::ColumnMajor, double, cutlass::layout::RowMajor, cutlass::arch::OpMultiplyAdd>::operator()(cutlass::Array<double, 4, true> &, const cutlass::Array<double, 4, true> &, const cutlass::Array<double, 2, true> &, const cutlass::Array<double, 4, true> &) const not implemented
void cutlass::arch::Mma<cutlass::gemm::GemmShape<16, 8, 8>, 32, double, cutlass::layout::RowMajor, double, cutlass::layout::ColumnMajor, double, cutlass::layout::RowMajor, cutlass::arch::OpMultiplyAdd>::operator()(cutlass::Array<double, 4, true> &, const cutlass::Array<double, 4, true> &, const cutlass::Array<double, 2, true> &, const cutlass::Array<double, 4, true> &) const not implemented
void cutlass::arch::Mma<cutlass::gemm::GemmShape<16, 8, 8>, 32, double, cutlass::layout::RowMajor, double, cutlass::layout::ColumnMajor, double, cutlass::layout::RowMajor, cutlass::arch::OpMultiplyAdd>::operator()(cutlass::Array<double, 4, true> &, const cutlass::Array<double, 4, true> &, const cutlass::Array<double, 2, true> &, const cutlass::Array<double, 4, true> &) const not implemented
void cutlass::arch::Mma<cutlass::gemm::GemmShape<16, 8, 8>, 32, double, cutlass::layout::RowMajor, double, cutlass::layout::ColumnMajor, double, cutlass::layout::RowMajor, cutlass::arch::OpMultiplyAdd>::operator()(cutlass::Array<double, 4, true> &, const cutlass::Array<double, 4, true> &, const cutlass::Array<double, 2, true> &, const cutlass::Array<double, 4, true> &) const not implemented
void cutlass::arch::Mma<cutlass::gemm::GemmShape<16, 8, 8>, 32, double, cutlass::layout::RowMajor, double, cutlass::layout::ColumnMajor, double, cutlass::layout::RowMajor, cutlass::arch::OpMultiplyAdd>::operator()(cutlass::Array<double, 4, true> &, const cutlass::Array<double, 4, true> &, const cutlass::Array<double, 2, true> &, const cutlass::Array<double, 4, true> &) const not implemented
void cutlass::arch::Mma<cutlass::gemm::GemmShape<16, 8, 8>, 32, double, cutlass::layout::RowMajor, double, cutlass::layout::ColumnMajor, double, cutlass::layout::RowMajor, cutlass::arch::OpMultiplyAdd>::operator()(cutlass::Array<double, 4, true> &, const cutlass::Array<double, 4, true> &, const cutlass::Array<double, 2, true> &, const cutlass::Array<double, 4, true> &) const not implemented
void cutlass::arch::Mma<cutlass::gemm::GemmShape<16, 8, 8>, 32, double, cutlass::layout::RowMajor, double, cutlass::layout::ColumnMajor, double, cutlass::layout::RowMajor, cutlass::arch::OpMultiplyAdd>::operator()(cutlass::Array<double, 4, true> &, const cutlass::Array<double, 4, true> &, const cutlass::Array<double, 2, true> &, const cutlass::Array<double, 4, true> &) const not implemented
void cutlass::arch::Mma<cutlass::gemm::GemmShape<16, 8, 8>, 32, double, cutlass::layout::RowMajor, double, cutlass::layout::ColumnMajor, double, cutlass::layout::RowMajor, cutlass::arch::OpMultiplyAdd>::operator()(cutlass::Array<double, 4, true> &, const cutlass::Array<double, 4, true> &, const cutlass::Array<double, 2, true> &, const cutlass::Array<double, 4, true> &) const not implemented
void cutlass::arch::Mma<cutlass::gemm::GemmShape<16, 8, 8>, 32, double, cutlass::layout::RowMajor, double, cutlass::layout::ColumnMajor, double, cutlass::layout::RowMajor, cutlass::arch::OpMultiplyAdd>::operator()(cutlass::Array<double, 4, true> &, const cutlass::Array<double, 4, true> &, const cutlass::Array<double, 2, true> &, const cutlass::Array<double, 4, true> &) const not implemented
void cutlass::arch::Mma<cutlass::gemm::GemmShape<16, 8, 8>, 32, double, cutlass::layout::RowMajor, double, cutlass::layout::ColumnMajor, double, cutlass::layout::RowMajor, cutlass::arch::OpMultiplyAdd>::operator()(cutlass::Array<double, 4, true> &, const cutlass::Array<double, 4, true> &, const cutlass::Array<double, 2, true> &, const cutlass::Array<double, 4, true> &) const not implemented
void cutlass::arch::Mma<cutlass::gemm::GemmShape<16, 8, 8>, 32, double, cutlass::layout::RowMajor, double, cutlass::layout::ColumnMajor, double, cutlass::layout::RowMajor, cutlass::arch::OpMultiplyAdd>::operator()(cutlass::Array<double, 4, true> &, const cutlass::Array<double, 4, true> &, const cutlass::Array<double, 2, true> &, const cutlass::Array<double, 4, true> &) const not implemented
void cutlass::arch::Mma<cutlass::gemm::GemmShape<16, 8, 8>, 32, double, cutlass::layout::RowMajor, double, cutlass::layout::ColumnMajor, double, cutlass::layout::RowMajor, cutlass::arch::OpMultiplyAdd>::operator()(cutlass::Array<double, 4, true> &, const cutlass::Array<double, 4, true> &, const cutlass::Array<double, 2, true> &, const cutlass::Array<double, 4, true> &) const not implemented
void cutlass::arch::Mma<cutlass::gemm::GemmShape<16, 8, 8>, 32, double, cutlass::layout::RowMajor, double, cutlass::layout::ColumnMajor, double, cutlass::layout::RowMajor, cutlass::arch::OpMultiplyAdd>::operator()(cutlass::Array<double, 4, true> &, const cutlass::Array<double, 4, true> &, const cutlass::Array<double, 2, true> &, const cutlass::Array<double, 4, true> &) const not implemented
void cutlass::arch::Mma<cutlass::gemm::GemmShape<16, 8, 8>, 32, double, cutlass::layout::RowMajor, double, cutlass::layout::ColumnMajor, double, cutlass::layout::RowMajor, cutlass::arch::OpMultiplyAdd>::operator()(cutlass::Array<double, 4, true> &, const cutlass::Array<double, 4, true> &, const cutlass::Array<double, 2, true> &, const cutlass::Array<double, 4, true> &) const not implemented
void cutlass::arch::Mma<cutlass::gemm::GemmShape<16, 8, 8>, 32, double, cutlass::layout::RowMajor, double, cutlass::layout::ColumnMajor, double, cutlass::layout::RowMajor, cutlass::arch::OpMultiplyAdd>::operator()(cutlass::Array<double, 4, true> &, const cutlass::Array<double, 4, true> &, const cutlass::Array<double, 2, true> &, const cutlass::Array<double, 4, true> &) const not implemented
void cutlass::arch::Mma<cutlass::gemm::GemmShape<16, 8, 8>, 32, double, cutlass::layout::RowMajor, double, cutlass::layout::ColumnMajor, double, cutlass::layout::RowMajor, cutlass::arch::OpMultiplyAdd>::operator()(cutlass::Array<double, 4, true> &, const cutlass::Array<double, 4, true> &, const cutlass::Array<double, 2, true> &, const cutlass::Array<double, 4, true> &) const not implemented
void cutlass::arch::Mma<cutlass::gemm::GemmShape<16, 8, 8>, 32, double, cutlass::layout::RowMajor, double, cutlass::layout::ColumnMajor, double, cutlass::layout::RowMajor, cutlass::arch::OpMultiplyAdd>::operator()(cutlass::Array<double, 4, true> &, const cutlass::Array<double, 4, true> &, const cutlass::Array<double, 2, true> &, const cutlass::Array<double, 4, true> &) const not implemented
void cutlass::arch::Mma<cutlass::gemm::GemmShape<16, 8, 8>, 32, double, cutlass::layout::RowMajor, double, cutlass::layout::ColumnMajor, double, cutlass::layout::RowMajor, cutlass::arch::OpMultiplyAdd>::operator()(cutlass::Array<double, 4, true> &, const cutlass::Array<double, 4, true> &, const cutlass::Array<double, 2, true> &, const cutlass::Array<double, 4, true> &) const not implemented
void cutlass::arch::Mma<cutlass::gemm::GemmShape<16, 8, 8>, 32, double, cutlass::layout::RowMajor, double, cutlass::layout::ColumnMajor, double, cutlass::layout::RowMajor, cutlass::arch::OpMultiplyAdd>::operator()(cutlass::Array<double, 4, true> &, const cutlass::Array<double, 4, true> &, const cutlass::Array<double, 2, true> &, const cutlass::Array<double, 4, true> &) const not implemented
void cutlass::arch::Mma<cutlass::gemm::GemmShape<16, 8, 8>, 32, double, cutlass::layout::RowMajor, double, cutlass::layout::ColumnMajor, double, cutlass::layout::RowMajor, cutlass::arch::OpMultiplyAdd>::operator()(cutlass::Array<double, 4, true> &, const cutlass::Array<double, 4, true> &, const cutlass::Array<double, 2, true> &, const cutlass::Array<double, 4, true> &) const not implemented
void cutlass::arch::Mma<cutlass::gemm::GemmShape<16, 8, 8>, 32, double, cutlass::layout::RowMajor, double, cutlass::layout::ColumnMajor, double, cutlass::layout::RowMajor, cutlass::arch::OpMultiplyAdd>::operator()(cutlass::Array<double, 4, true> &, const cutlass::Array<double, 4, true> &, const cutlass::Array<double, 2, true> &, const cutlass::Array<double, 4, true> &) const not implemented
void cutlass::arch::Mma<cutlass::gemm::GemmShape<16, 8, 8>, 32, double, cutlass::layout::RowMajor, double, cutlass::layout::ColumnMajor, double, cutlass::layout::RowMajor, cutlass::arch::OpMultiplyAdd>::operator()(cutlass::Array<double, 4, true> &, const cutlass::Array<double, 4, true> &, const cutlass::Array<double, 2, true> &, const cutlass::Array<double, 4, true> &) const not implemented
void cutlass::arch::Mma<cutlass::gemm::GemmShape<16, 8, 8>, 32, double, cutlass::layout::RowMajor, double, cutlass::layout::ColumnMajor, double, cutlass::layout::RowMajor, cutlass::arch::OpMultiplyAdd>::operator()(cutlass::Array<double, 4, true> &, const cutlass::Array<double, 4, true> &, const cutlass::Array<double, 2, true> &, const cutlass::Array<double, 4, true> &) const not implemented
void cutlass::arch::Mma<cutlass::gemm::GemmShape<16, 8, 8>, 32, double, cutlass::layout::RowMajor, double, cutlass::layout::ColumnMajor, double, cutlass::layout::RowMajor, cutlass::arch::OpMultiplyAdd>::operator()(cutlass::Array<double, 4, true> &, const cutlass::Array<double, 4, true> &, const cutlass::Array<double, 2, true> &, const cutlass::Array<double, 4, true> &) const not implemented
void cutlass::arch::Mma<cutlass::gemm::GemmShape<16, 8, 8>, 32, double, cutlass::layout::RowMajor, double, cutlass::layout::ColumnMajor, double, cutlass::layout::RowMajor, cutlass::arch::OpMultiplyAdd>::operator()(cutlass::Array<double, 4, true> &, const cutlass::Array<double, 4, true> &, const cutlass::Array<double, 2, true> &, const cutlass::Array<double, 4, true> &) const not implemented
void cutlass::arch::Mma<cutlass::gemm::GemmShape<16, 8, 8>, 32, double, cutlass::layout::RowMajor, double, cutlass::layout::ColumnMajor, double, cutlass::layout::RowMajor, cutlass::arch::OpMultiplyAdd>::operator()(cutlass::Array<double, 4, true> &, const cutlass::Array<double, 4, true> &, const cutlass::Array<double, 2, true> &, const cutlass::Array<double, 4, true> &) const not implemented
Failed to synchronize device after kernel launch.
```

I am wondering if there are such PTX instructions, why I cannot set the instruction shapes other than the default `m8n8k4` ? And why does this `DefaultMmaTensorOp` function not need a `SM_arch` parameter?

**Q2.**
As I also see in other examples, such as `example_14_ampere_tf32_tensorop_gemm`, it defines `Gemm` as follows:

```
using ShapeMMAThreadBlock =
    cutlass::gemm::GemmShape<128, 128, 16>;  // <- threadblock tile M = 128, N = 128, K = 16
// This code section describes tile size a warp will compute
using ShapeMMAWarp = cutlass::gemm::GemmShape<64, 64, 16>;  // <- warp tile M = 64, N = 64, K = 16
// This code section describes the size of MMA op
using ShapeMMAOp = cutlass::gemm::GemmShape<16, 8, 8>;  // <- MMA Op tile M = 16, N = 8, K = 8

// This code section describes how threadblocks are scheduled on GPU
using SwizzleThreadBlock = cutlass::gemm::threadblock::GemmIdentityThreadblockSwizzle<>;  // <- ??

// This code section describes the epilogue part of the kernel
using EpilogueOp = cutlass::epilogue::thread::LinearCombination<
    ElementOutput,                                     // <- data type of output matrix
    128 / cutlass::sizeof_bits<ElementOutput>::value,  // <- the number of elements per vectorized
                                                       // memory access. For a byte, it's 16
                                                       // elements. This becomes the vector width of
                                                       // math instructions in the epilogue too
    ElementAccumulator,                                // <- data type of accumulator
    ElementComputeEpilogue>;  // <- data type for alpha/beta in linear combination function

// Number of pipelines you want to use
constexpr int NumStages = 4;

using Gemm = cutlass::gemm::device::Gemm<ElementInputA,
                                         LayoutInputA,
                                         ElementInputB,
                                         LayoutInputB,
                                         ElementOutput,
                                         LayoutOutput,
                                         ElementAccumulator,
                                         MMAOp,
                                         SmArch,
                                         ShapeMMAThreadBlock,
                                         ShapeMMAWarp,
                                         ShapeMMAOp,
                                         EpilogueOp,
                                         SwizzleThreadBlock,
                                         NumStages>;

```
 Here, I also tried to change the `MMA` op shape from `using ShapeMMAOp = cutlass::gemm::GemmShape<16, 8, 8>` into `using ShapeMMAOp = cutlass::gemm::GemmShape<16, 8, 16>`, but again, it raises some errors.

My general question is what is the correct way to determine the MMA instruction shape? 

Thanks!

**Comment by hwu36** [comment]
as to your q1. 884 is the only one available on a100

**Comment by hwu36** [comment]
on ampere, tf32 is using 1688.  fp16/bf16 is using 16816.  s8/u8 is using 16832.  s4/us is using 16864.  b1 is using 168256.

**Comment by YJHMITWEB** [comment]
@hwu36 Thanks for answering. If I understand correctly, it means the more bits the element type has, the fewer elements the tensor core can process. To this, I have an extended question:

In `example_14_ampere_tf32_tensorop_gemm`, I changed the datatype from `float` to `cutlass::half_t`, and I found `m16n8k8` can still work, does this mean that there are some computing power wasted in each warp MMA? As it can do `m16n8k16`, but I only give it `m16n8k8`.
